lib/TPP/Transforms/CombineXsmmPass.cpp

```diff
@@ -0,0 +1,179 @@
+//===CombineXsmmPass.cpp --------------------------------------*----C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM
+// Exceptions. / See https://llvm.org/LICENSE.txt for license information. /
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+
+#include "TPP/Dialect/Xsmm/XsmmOps.h"
+#include "TPP/Dialect/Xsmm/XsmmUtils.h"
+#include "TPP/Passes.h"
+#include "TPP/Transforms/Utils/BuilderUtils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+using namespace mlir;
+
+#include "TPP/Dialect/Xsmm/XsmmUtils.h"
+#include "TPP/Transforms/Utils/VNNIUtils.h"
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DEF_COMBINEXSMMOPPASS
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+namespace {
+
+static FailureOr<DenseI64ArrayAttr>
+getSizesAndLeadingDimForBrgemmOp(RewriterBase &rewriter, xsmm::BrgemmOp opTy) {
+
+  auto memrefA = opTy.getOperand(1).getType();
+  auto memrefB = opTy.getOperand(2).getType();
+  auto memrefC = opTy.getOperand(3).getType();
+
+  if (!isa<ShapedType>(memrefC) || !isa<ShapedType>(memrefA)) {
+    return failure();
+  }
+  int64_t m, n, k;
+  m = memrefC.cast<ShapedType>().getShape()[0];
+  n = memrefC.cast<ShapedType>().getShape()[1];
+  k = memrefA.cast<ShapedType>().getShape()[2];
+
+  auto ldaDim = xsmm::utils::getLeadingDim(memrefA, /*pos=*/1);
+  if (failed(ldaDim)) {
+    return failure();
+  }
+  int64_t lda = *ldaDim;
+
+  auto ldbDim = xsmm::utils::getLeadingDim(memrefB, /*pos=*/1);
+  if (failed(ldbDim)) {
+    return failure();
+  }
+  int64_t ldb =
+      (vnni::utils::isInVnniLayout(vnni::utils::VnniOperandRank::BRGEMM_INS,
+                                   memrefB.cast<MemRefType>()))
+          ? *ldbDim / *vnni::utils::getVnniBlockingFactor(memrefB)
+          : *ldbDim;
+
+  auto ldcDim = xsmm::utils::getLeadingDim(memrefC);
+  if (failed(ldcDim)) {
+    return failure();
+  }
+  int64_t ldc = *ldcDim;
+
+  // If we are dealing with a BRGEMM we need to pass two extra dimensions:
+  // - strideA and strideB that represent the stride between different GEMM
+  // in BRGEMM.
+  int64_t strideA = lda * m;
+  int64_t strideB = ldb * k;
+  return DenseI64ArrayAttr::get(
+      rewriter.getContext(),
+      ArrayRef<int64_t>{m, n, k, lda, ldb, ldc, strideA, strideB});
+}
+
+static ArrayAttr getBrgemmFlags(RewriterBase &rewriter, xsmm::BrgemmOp opTy) {
+  auto memrefB = opTy.getOperand(2).getType().cast<MemRefType>();
+  xsmm::GemmFlagsAttr gemmFlag =
```

Comment:
What if we have BETA_0, do we loose it? This method does not looks correct to me.

Reply:
Ah, yes! @alheinecke @egeor does libxsmm support fused brgemm with BETA=0?

Reply:
My interpretation of libxsmm's code is that the `LIBXSMM_GEMM_FLAG_BETA_0` is handled on all tile/packed/fused generators (mostly on store sequences), so it's independent of which op was originally called.

The libxsmm-dnn example for DNN models seems to corroborate it: https://github.com/libxsmm/libxsmm-dnn/blob/2075b99f38af245a81293de1d0b35a5224de93c7/src/libxsmm_dnn_fc.c#L289

Anyway, if we don't detect BETA=0 we won't do any fusion in real world code because both TF and PT examples get BETA=0 through out code. You should pass the flag and make sure the execution is correct in the integration test, as below.

Reply:
You are now ignoring BETA=0, which means it will not apply to any current integration test or benchmark. In addition, your integration test does not check the output of the execution, which means we don't know if it's doing the right thing in addition to lowering correctly.

You also seem to have removed the `xsmm.zero` from the IR tests below, which means now you have adapted your tests to the code and not the other way around. This way, we'll never know if your code can match the IR we actually see in the wild.

Reply:
I thought beta=0 was a case which was no longer supposed to be fused, therefore I am not sure how to make it applicable to an integration test other than build a case where I initialize beta to 0 and not fuse. As for the end to end case, I will update the integration test to check results. 

I am not sure what you mean by xsmm.zero, do you mean beta assigned to xsmm.zero?

Reply:
> I thought beta=0 was a case which was no longer supposed to be fused

I thought we could (as per my now deprecated comment above), but later I realised (and commented above) that all BRGEMMs we have on tests and benchmarks have BETA=0, so we cannot ignore it.

> I am not sure what you mean by xsmm.zero, do you mean beta assigned to xsmm.zero?

Sorry, I meant `xsmm.brgemm.dispatch ... flags = (beta_0)` to `flags = (none)` in https://github.com/plaidml/tpp-mlir/pull/752/commits/e2a3597d066c410d2f0e52214404181638652c50.